core/src/ldml/ldml_processor.cpp

```diff
@@ -336,5 +352,76 @@ km_kbp_status ldml_processor::validate() const {
   return _valid ? KM_KBP_STATUS_OK : KM_KBP_STATUS_INVALID_KEYBOARD;
 }
 
+void
+ldml_processor::emit_text(km_kbp_state *state, const std::u16string &str) {
+  const std::u32string str32 = kmx::u16string_to_u32string(str);
+  emit_text(state, str32);
+}
+
+void
+ldml_processor::emit_text(km_kbp_state *state, const std::u32string &str) {
+  for (auto it = str.begin(); it < str.end(); it++) {
+    const auto ch = *it;
+    // If we are at the start of a sequence:
+    if (ch == LDML_UC_SENTINEL) {
+      it++; // consume LDML_UC_SENTINEL
+      // TODO-LDML: Might assert if a malformed sequence is included- "should not happen"?
+      assert(it < str.end());
+      // verify that the next char is LDML_MARKER_CODE
+      assert(*it == LDML_MARKER_CODE);
+      it++; // consume LDML_MARKER_CODE
+      assert(it < str.end());
+      const auto marker_no = *it;
+      assert(marker_no >= LDML_MARKER_MIN_INDEX);
+      assert(marker_no <= LDML_MARKER_ANY_INDEX);
+      emit_marker(state, marker_no);
+    } else {
+      emit_text(state, ch);
+    }
+  }
+}
+
+void
+ldml_processor::emit_text(km_kbp_state *state, km_kbp_usv ch) {
+  assert(ch != LDML_UC_SENTINEL);
```

Comment:
This is more of an internal assert to make sure that the logic isn't misfiring as if someone made an API error and called `emit_text(keys[n].to.value)` to emit a key directly.

we could separately have:
- a pre-scan of all key=to and transform=to strings and make sure there's no illegal unicode - this would be a validation error, not an assert. This would be at runtime.   kmc should also do the same so we catch it at compile time 
- we could retain an assert on no illegal unicode _here_, that would make sure the upcoming regex machinery doesn't emit issues. But this wouldn't be at production runtime. 

Seems like a separate ticket to me? it's drifted from the mark(er) consideration.

Reply:
filed https://github.com/keymanapp/keyman/issues/9446